### PR TITLE
[wip]ddl: refactor the args for the DDL type of renametable

### PR DESCRIPTION
**Pull request by joccau**
### What problem does this PR solve?

Issue Number: Ref https://github.com/pingcap/tidb/issues/53930
 Ref https://github.com/pingcap/tidb/issues/54436

Problem Summary:
- Refactor the args in ddl job for the type of `renametable`

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/meta/model/job_args.go [hunk 1]
```diff
@@ -109,3 +110,98 @@ func GetTruncateTableArgsAfterRun(job *Job) (*TruncateTableArgs, error) {
...
+	}
+	return argsV2, err
+}
+
+// GetRenameTableUniqueIDs gets unique IDs from job for rename table type.
+func GetRenameTableUniqueIDs(job *Job, schema bool) []int64 {
+	args := &RenameTableArgs{}
+
+	if job.Version == JobVersion1 {
+		args.SchemaIDs = job.CtxVars[0].([]int64)
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:197]: we don't need to use `CtxVars` now even for version 1, just store it in the `args` without persist them

we don't need to write a exported method for a only used-once logic

and i found a better way to pass those args during submitting, see https://github.com/pingcap/tidb/pull/55964/files#diff-7328439fbfe5c8d8a925b3de0a38dd6393f097a8015e448d5f8ec1a0e456fe24R156-R157

File: pkg/ddl/ddl_test.go [hunk 2]
```diff
@@ -202,6 +202,18 @@ func TestIgnorableSpec(t *testing.T) {
 }
 
 func TestBuildJobDependence(t *testing.T) {
+	vers := []model.JobVersion{
+		model.JobVersion1,
+		model.JobVersion2,
+	}
+
+	for _, ver := range vers {
+		model.SetJobVerInUse(ver)
```
Comment (D3Hunter) [inline pkg/ddl/ddl_test.go:211]: by setting global `SetJobVerInUse`, you must revert back on finish, it's more simple to pass the version directly into `testBuildJobDependence`.

Comment (D3Hunter) [reply pkg/ddl/ddl_test.go:211]: those test is for DDL job queue which only works in version below 6.2.0, i think it's ok to leave them in job version1

**Comment by D3Hunter** [comment]
link this issue https://github.com/pingcap/tidb/issues/53930

File: pkg/meta/model/job_args.go [hunk 3]
```diff
@@ -109,3 +110,98 @@ func GetTruncateTableArgsAfterRun(job *Job) (*TruncateTableArgs, error) {
...
+	// for Args
+	OldSchemaID  int64        `json:"old_schema_id,omitempty"`
+	SchemaName   pmodel.CIStr `json:"schema_name,omitempty"`
+	NewTableName pmodel.CIStr `json:"new_table_name,omitempty"`
+
+	// for CtxVars
+	// SchemaIDs contains {oldSchemaID, newSchemaID}
+	SchemaIDs []int64 `json:"-"`
+	// TableIDs contains {tableID}
+	TableIDs []int64 `json:"-"`
```
[the remark this answers is not in the document]
Comment (joccau) [reply pkg/meta/model/job_args.go:273]: > the job already have `TableID`

The job has already NewSchemaID in job.SchemaID. maybe remove all.

File: pkg/meta/model/job_args.go [hunk 4]
```diff
@@ -257,3 +258,87 @@ func getTruncateTableArgs(job *Job, argsOfFinished bool) (*TruncateTableArgs, er
...
+	argsV2, err := getOrDecodeArgsV2[*RenameTableArgs](job)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return argsV2, nil
+}
+
+// UpdateRenameTableArgs updates the rename table args.
+// need to reset the old schema ID to new schema ID.
+func UpdateRenameTableArgs(job *Job) error {
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:318]: use fillFinishedJob

File: pkg/meta/model/job_args.go [hunk 5]
```diff
@@ -257,3 +258,87 @@ func getTruncateTableArgs(job *Job, argsOfFinished bool) (*TruncateTableArgs, er
...
+		}
+		return &args, nil
+	}
+
+	// for version V2
+	argsV2, err := getOrDecodeArgsV2[*RenameTableArgs](job)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return argsV2, nil
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:313]:
```suggestion
	return getOrDecodeArgsV2[*RenameTableArgs](job)
```

File: pkg/meta/model/job_args.go [hunk 6]
```diff
@@ -257,3 +258,87 @@ func getTruncateTableArgs(job *Job, argsOfFinished bool) (*TruncateTableArgs, er
...
+
+// GetRenameTableArgs get the arguments from job.
+func GetRenameTableArgs(job *Job) (*RenameTableArgs, error) {
+	var (
+		oldSchemaID   int64
+		oldSchemaName pmodel.CIStr
+		newTableName  pmodel.CIStr
+	)
+
+	if job.Version <= JobVersion1 {
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:286]: no need check version `< 1`

File: pkg/ddl/table_test.go [hunk 7]
```diff
@@ -76,6 +83,7 @@ func testRenameTable(
 func testRenameTables(t *testing.T, ctx sessionctx.Context, d ddl.ExecutorForTest, oldSchemaIDs, newSchemaIDs []int64, newTableNames []*pmodel.CIStr, oldTableIDs []int64, oldSchemaNames, oldTableNames []*pmodel.CIStr) *model.Job {
 	job := &model.Job{
 		Type:       model.ActionRenameTables,
+		Version:    model.GetJobVerInUse(),
```
Comment (D3Hunter) [inline pkg/ddl/table_test.go:86]: this pr didn't handle ActionRenameTables, it won't work for v2

File: pkg/meta/model/job_args.go (hunk 4 above)
Comment (D3Hunter) [reply pkg/meta/model/job_args.go:318]: seems not needed to update it, i will check it later

File: pkg/ddl/executor.go [hunk 8]
```diff
@@ -6435,6 +6441,21 @@ func (e *executor) DoDDLJobWrapper(ctx sessionctx.Context, jobW *JobWrapper) (re
...
 }
 
+func getRenameTableUniqueIDs(job *model.Job, schema bool) []int64 {
+	if !schema {
+		return []int64{job.TableID}
+	}
+
+	var oldSchemaID int64
+	if job.Version <= model.JobVersion1 {
+		oldSchemaID = job.Args[0].(int64)
```
Comment (D3Hunter) [inline pkg/ddl/executor.go:6451]: the first v1 args is a slice

Comment (D3Hunter) [reply pkg/ddl/executor.go:6451]: psss it by JobWrapper.JobArgs, so we can directly access it here

Comment (joccau) [reply pkg/ddl/executor.go]: Done

**Review by lance6716 — APPROVED** [review]

File: pkg/ddl/executor.go [hunk 9]
```diff
@@ -4268,18 +4268,24 @@ func (e *executor) renameTable(ctx sessionctx.Context, oldIdent, newIdent ast.Id
...
 		SQLMode: ctx.GetSessionVars().SQLMode,
 	}
 
-	err = e.DoDDLJob(ctx, job)
+	args := &model.RenameTableArgs{
+		OldSchemaID:   schemas[0].ID,
+		OldSchemaName: schemas[0].Name,
+		NewTableName:  newIdent.Name,
+	}
+	job.FillArgs(args)
```
Comment (D3Hunter) [inline pkg/ddl/executor.go:4286]:
```suggestion
```

File: pkg/ddl/executor.go [hunk 10]
```diff
@@ -6269,6 +6275,9 @@ func (e *executor) DoDDLJobWrapper(ctx sessionctx.Context, jobW *JobWrapper) (re
 		// Instead, we merge all the jobs into one pending job.
 		return appendToSubJobs(mci, job)
 	}
+
+	logutil.DDLLogger().Info("Debug ddl job", zap.Bool("jobArgs is nil", jobW.JobArgs == nil))
+
```
Comment (D3Hunter) [inline pkg/ddl/executor.go:6280]:
```suggestion
```

File: pkg/ddl/job_scheduler.go [hunk 11]
```diff
@@ -680,6 +680,9 @@ func insertDDLJobs2Table(ctx context.Context, se *sess.Session, jobWs ...*JobWra
 			jobW.FillArgs(jobW.JobArgs)
 		}
 		injectModifyJobArgFailPoint(jobWs)
+
+		logutil.DDLLogger().Info("Debug ddl job", zap.Bool("jobArgs is nil", jobW.JobArgs == nil))
```
Comment (D3Hunter) [inline pkg/ddl/job_scheduler.go:684]: remove it too

File: pkg/ddl/table_test.go [hunk 12]
```diff
@@ -52,17 +52,24 @@ func testRenameTable(
...
 			{Database: oldSchemaName.L, Table: tblInfo.Name.L},
 			{Database: newSchemaName.L, Table: tblInfo.Name.L},
 		},
 	}
+	args := model.RenameTableArgs{
+		OldSchemaID:   oldSchemaID,
+		OldSchemaName: oldSchemaName,
+		NewTableName:  tblInfo.Name,
+	}
+	job.FillArgs(&args)
```
Comment (D3Hunter) [inline pkg/ddl/table_test.go:71]: not needed now, we can pass it directly to DoDDLJobWrapper

File: pkg/meta/model/job_args_test.go [hunk 13]
```diff
@@ -226,3 +227,75 @@ func TestTruncateTableArgs(t *testing.T) {
...
+		job := &Job{
+			Version: jobver,
+			Type:    ActionRenameTable,
+		}
+		job.FillArgs(inArgs)
+
+		// encode job
+		_, err := job.Encode(true)
+		require.NoError(t, err)
+		job.Args = nil
```
Comment (D3Hunter) [inline pkg/meta/model/job_args_test.go:249]: use getJobBytes

File: pkg/meta/model/job_args_test.go [hunk 14]
```diff
@@ -226,3 +227,75 @@ func TestTruncateTableArgs(t *testing.T) {
...
+
+		// get the args after decode
+		args, err := GetRenameTableArgs(job)
+		require.NoError(t, err)
+		require.Equal(t, args, inArgs)
```
Comment (D3Hunter) [inline pkg/meta/model/job_args_test.go:254]:
```suggestion
		require.Equal(t, inArgs, args)
```
the first param is `expected`

File: pkg/meta/model/job_args_test.go [hunk 15]
```diff
@@ -226,3 +227,75 @@ func TestTruncateTableArgs(t *testing.T) {
...
+
+		// get the args from args
+		args, err = GetRenameTableArgs(job)
+		require.NoError(t, err)
+		require.Equal(t, args, inArgs)
```
Comment (D3Hunter) [inline pkg/meta/model/job_args_test.go:259]: doitto

File: pkg/meta/model/job_args_test.go [hunk 16]
```diff
@@ -226,3 +227,75 @@ func TestTruncateTableArgs(t *testing.T) {
...
+		})
+
+		// get args from job.RawArgs
+		job.Args = nil
+		args, err = GetRenameTableArgs(job)
+		require.NoError(t, err)
+		require.Equal(t, args, &RenameTableArgs{
+			OldSchemaID:   9528,
+			OldSchemaName: model.NewCIStr("old_schema_name"),
+			NewTableName:  model.NewCIStr("new_table_name"),
```
Comment (D3Hunter) [inline pkg/meta/model/job_args_test.go:298]: doittoo

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/ddl/executor.go (hunk 10 above)
Comment (joccau) [reply pkg/ddl/executor.go]: done

File: pkg/ddl/job_scheduler.go (hunk 11 above)
Comment (joccau) [reply pkg/ddl/job_scheduler.go]: done

File: pkg/ddl/table_test.go (hunk 12 above)
Comment (joccau) [reply pkg/ddl/table_test.go]: done

File: pkg/meta/model/job_args_test.go (hunk 13 above)
Comment (joccau) [reply pkg/meta/model/job_args_test.go]: done

File: pkg/meta/model/job_args_test.go (hunk 14 above)
Comment (joccau) [reply pkg/meta/model/job_args_test.go]: done

File: pkg/meta/model/job_args_test.go (hunk 15 above)
Comment (joccau) [reply pkg/meta/model/job_args_test.go]: done

File: pkg/meta/model/job_args_test.go (hunk 16 above)
Comment (joccau) [reply pkg/meta/model/job_args_test.go]: done

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**